controller/src/lib.rs

```diff
@@ -20,114 +25,204 @@ pub mod plan;
-
-            if let Some(backend) = locked {
-                tracing::info!(lock=%lock, "Lock is held.");
-
-                // Anything that fails to find the drone results in an error here, since we just
-                // checked that the lock is held which implies that the drone exists.
-
-                let (drone, bearer_token) = {
-                    let state = state.state();
-                    let cluster = state
-                        .cluster(&schedule_request.value.cluster)
-                        .ok_or_else(|| anyhow!("Expected cluster to exist."))?;
-
-                    let backend_state = cluster
-                        .backend(&backend)
-                        .ok_or_else(|| anyhow!("Lock held by a backend that doesn't exist."))?;
+async fn spawn_backend(
+    ref nats: TypedNats,
+    drone: DroneId,
+    schedule_request: &ScheduleRequest,
+) -> anyhow::Result<(ScheduleResponse, Option<u64>)> {
+    let timer = Timer::new();
+    let spawn_request = schedule_request.schedule(&drone);
+    match nats.request(&spawn_request).await {
+        Ok(true) => {
+            tracing::info!(
+                duration=?timer.duration(),
+                backend_id=%spawn_request.backend_id,
+                %drone,
+                "Drone accepted backend."
+            );
+
+            let seq_id = nats
+                .publish_jetstream(&WorldStateMessage {
+                    cluster: schedule_request.cluster.clone(),
+                    message: ClusterStateMessage::BackendMessage(BackendMessage {
+                        backend: spawn_request.backend_id.clone(),
+                        message: BackendMessageType::Assignment {
+                            drone: drone.clone(),
+                            lock: schedule_request.lock.clone(),
+                            bearer_token: spawn_request.bearer_token.clone(),
+                        },
+                    }),
+                })
+                .await?;
+
+            tracing::info!(
+            logical_time=?seq_id,
+            "backend state updated at time"
+                );
+
+            Ok((
+                ScheduleResponse::Scheduled {
+                    drone,
+                    backend_id: spawn_request.backend_id,
+                    bearer_token: spawn_request.bearer_token.clone(),
+                    spawned: true,
+                },
+                Some(seq_id),
+            ))
+        }
+        Ok(false) => {
+            tracing::warn!("Drone rejected backend.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+        Err(error) => {
+            tracing::warn!(?error, "Scheduler returned error.");
+            Ok((ScheduleResponse::NoDroneAvailable, None))
+        }
+    }
+}
 
-                    let drone = backend_state.drone.clone().ok_or_else(|| {
-                        anyhow!("Lock held by a backend without a drone assignment.")
-                    })?;
+fn locked_backend(
```

Comment:
This could be more descriptive, maybe `backend_of_lock` or add a docstring to that effect?